flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunction.scala

```diff
@@ -131,7 +131,8 @@ class GroupAggProcessFunction(
 
       // if this was not the first row and we have to emit retractions
       if (generateRetraction && !firstRow) {
-        if (prevRow.row.equals(newRow.row)) {
+        // the condition of !stateCleaningEnabled is avoided state to be cleaned up too early
```

Comment:
I would move the description into the `if` block and modify it as follows:
```
// newRow is the same as before and state cleaning is not enabled. We do not emit retraction and acc message.
// If state cleaning is enabled, we have to emit messages to prevent too early state eviction of downstream operators.
```